File: sdks/csharp/src/Synap.SDK/Transport.cs
using System.Buffers.Binary;
using System.Collections.Concurrent;
using System.Net.Sockets;
using System.Text;
using HiveLLM.Thunder;
using Synap.SDK.Exceptions;

namespace Synap.SDK;

/// <summary>Transport mode for the Synap client.</summary>
public enum TransportMode
{
    /// <summary>SynapRPC native protocol over TCP with MessagePack (default, port 15501).</summary>
    SynapRpc,

    /// <summary>RESP3 Redis-compatible protocol over TCP (port 6379).</summary>
    Resp3,

    /// <summary>HTTP/StreamableHTTP protocol (port 15500).</summary>
    Http,
}

// ---------------------------------------------------------------------------
// Wire value helpers for SynapRPC
// ---------------------------------------------------------------------------
internal static class WireValue
{
    /// <summary>
    /// Converts a plain CLR value into a Thunder <see cref="Value"/>.
    /// </summary>
    /// <remarks>
    /// The externally-tagged encoding this used to build by hand is Thunder's
    /// job now; what stays here is Synap's mapping from the CLR types its
    /// command mappers produce.
    /// </remarks>
    internal static Value ToWire(object? v) => v switch
    {
        null => Value.Null,
        bool b => Value.Bool(b),
        string s => Value.Str(s),
        byte[] bytes => Value.Bytes(bytes),
        long l => Value.Int(l),
        int i => Value.Int(i),
        double d => Value.Float(d),
        float f => Value.Float(f),
        _ => Value.Str(v.ToString() ?? string.Empty),
    };

    /// <summary>
    /// Converts a Thunder <see cref="Value"/> back to a plain CLR value.
    /// </summary>
    /// <remarks>
    /// <c>Bytes</c> decode to <see cref="string"/> when they are valid UTF-8 and
    /// stay <c>byte[]</c> otherwise. Thunder handles both the canonical
    /// MessagePack <c>bin</c> form the server emits from 1.1.0 and the legacy
    /// array-of-integers form, so a pre-1.1.0 server still interoperates.
    /// </remarks>
    internal static object? FromWire(Value? wire)
    {
        if (wire is null || wire.Kind == ValueKind.Null)
        {
            return null;
        }

        return wire.Kind switch
        {
            ValueKind.Str => wire.AsStr(),
            ValueKind.Bool => wire.AsBool(),
            ValueKind.Int => wire.AsInt(),
            ValueKind.Float => wire.AsFloat(),
            ValueKind.Bytes => DecodeBytes(wire.AsBytes()),
            ValueKind.Array => wire.AsArray()?.Select(item => FromWire(item)).ToList(),
            ValueKind.Map => wire.AsMap()?.ToDictionary(
                pair => FromWire(pair.Key)?.ToString() ?? string.Empty,
                pair => FromWire(pair.Value),
                StringComparer.Ordinal),
            _ => null,
        };
    }

    /// <summary>UTF-8 bytes surface as a string; anything else stays binary.</summary>
    private static object? DecodeBytes(byte[]? bytes)
    {
        if (bytes is null)
        {
            return null;
        }

        try
        {
            return new UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true)
                .GetString(bytes);
        }
        catch (ArgumentException)
        {
            return bytes;
        }
    }
}

// ---------------------------------------------------------------------------
// Command mapper — translates dot-notation ops to native protocol commands
// ---------------------------------------------------------------------------

File: sdks/php/src/Transport.php
<?php

declare(strict_types=1);

namespace Synap\SDK;

/**
 * Transport mode constants.
 */
final class TransportMode
{
    /** SynapRPC binary protocol over TCP (default). Port 15501. */
    public const SYNAP_RPC = 'synaprpc';

    /** RESP3 Redis-compatible text protocol over TCP. Port 6379. */
    public const RESP3 = 'resp3';

    /** HTTP / StreamableHTTP fallback. Port 15500. */
    public const HTTP = 'http';
}

// ── Wire value helpers ─────────────────────────────────────────────────────────

/**
 * Wrap a PHP value in the externally-tagged WireValue envelope (rmp_serde format).
 *
 * Supported variants:
 *   null    → "Null"
 *   bool    → {"Bool": bool}
 *   int     → {"Int": int}
 *   float   → {"Float": float}
 *   string  → {"Str": string}
 *
 * @param mixed $v
 * @return mixed
 */
function toWireValue(mixed $v): mixed
{
    if ($v === null) {
        return 'Null';
    }
    if (is_bool($v)) {
        return ['Bool' => $v];
    }
    if (is_int($v)) {
        return ['Int' => $v];
    }
    if (is_float($v)) {
        return ['Float' => $v];
    }
    if (is_string($v)) {
        return ['Str' => $v];
    }
    // Fallback: stringify any other scalar
    return ['Str' => (string) $v];
}

/**
 * Unwrap a WireValue envelope back to a plain PHP value.
 *
 * Handles all externally-tagged variants produced by the Synap server:
 *   "Null"           → null
 *   {"Str": "x"}     → string
 *   {"Int": 42}      → int
 *   {"Float": 3.14}  → float
 *   {"Bool": true}   → bool
 *   {"Bytes": ...}   → raw bytes (string)
 *   {"Array": [...]} → list<mixed>
 *   {"Map": [[k,v]]} → assoc array
 *
 * @param mixed $wire
 * @return mixed
 */
function fromWireValue(mixed $wire): mixed
{
    if ($wire === 'Null' || $wire === null) {
        return null;
    }
    if (is_array($wire)) {
        if (isset($wire['Str'])) {
            return $wire['Str'];
        }
        if (isset($wire['Int'])) {
            return $wire['Int'];
        }
        if (isset($wire['Float'])) {
            return $wire['Float'];
        }
        if (isset($wire['Bool'])) {
            return $wire['Bool'];
        }
        if (isset($wire['Bytes'])) {
            return $wire['Bytes'];
        }
        if (isset($wire['Array'])) {
            return array_map(__NAMESPACE__ . '\\fromWireValue', $wire['Array']);
        }
        if (isset($wire['Map'])) {
            $result = [];
            foreach ($wire['Map'] as [$k, $v]) {
                $result[(string) fromWireValue($k)] = fromWireValue($v);
            }
            return $result;
        }
    }
    return $wire;
}
